design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,238 @@
+
+	// StartTimestamp records the time a backup was started.
+	// Separate from CreationTimestamp, since that value changes
+	// on restores.
+	// The server's time is used for StartTimestamps
+	// +optional
+	// +nullable
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+
+	// CompletionTimestamp records the time a backup was completed.
+	// Completion time is recorded even on failed backups.
+	// Completion time is recorded before uploading the backup object.
+	// The server's time is used for CompletionTimestamps
+	// +optional
+	// +nullable
+	CompletionTimestamp *metav1.Time `json:"completionTimestamp,omitempty"`
+
+	// Progress holds the total number of bytes of the volume and the current
+	// number of backed up bytes. This can be used to display progress information
+	// about the backup operation.
+	// +optional
+	Progress VolumeOperationProgress `json:"progress,omitempty"`
+}
+
+type VolumeOperationProgress struct {
+	TotalBytes int64
+	BytesDone int64
+}
+
+type VolumePluginBackup struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec VolumePluginBackupSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status VolumePluginBackupStatus `json:"status,omitempty"`
+}
+```
+
+For every backup operation of volume, volume snapshotter creates VolumePluginBackup CR in Velero namespace.
+It keep updating the progress of operation along with other details like Volume name, Backup Name, SnapshotID etc as mentioned in the CR.
+
+In order to know the CR created for the particular backup of a volume, volume snapshotters adds following labels to CR:
+`velero.io/backup-name` with value as Backup Name, and,
+`velero.io/volume-name` with value as volume that is undergoing backup
+Backup name being unique won't cause issues like duplicates in identifying the CR.
+
+Though no restrictions are required on the name of CR, as a general practice, volume snapshotter can name this CR with the value same as return value of CreateSnapshot.
+
+After return from `CreateSnapshot` in `takePVSnapshot`, if VolumePluginBackup CR exists for particular backup of the volume, velero adds this CR to `backupRequest`.
+During persistBackup call, this CR also will be backed up to backup location.
+
+In backupSyncController, it checks for any VolumePluginBackup CRs that need to be synced from backup location, and syncs them to cluster if needed.
+
+### VolumePluginRestore CR
+
+```
+// VolumePluginRestoreSpec is the specification for a VolumePluginRestore CR.
+type VolumePluginRestoreSpec struct {
+	// SnapshotID is the identifier for the snapshot of the volume.
+	// This will be used to relate with output in 'velero describe backup'
+	SnapshotID string `json:"snapshotID"`
+
+	// Backup name
+	Backup string `json:"backup"`
+
+	// Plugin name
+	Plugin string `json:"plugin"`
+
+	// PluginSpecific are a map of key-value pairs that plugin want to provide
```

Comment:
Can you give an example of what values might go in here, and how it might differ from [ConfigMaps that we use](https://velero.io/docs/v1.4/custom-plugins/#plugin-configuration)?

Reply:
I hadn't thought any specific, but, added a way so that plugins can provide more info specific to this restore operation to users.
Configmaps in plugin is related to config at plugin level. This field is specific to details regarding this restore operation.

Reply:
One thing which I had in my mind when added this was specific to OpenEBS plugin, but, also thought that every plugin may have their specific things to convey to user about that particular backup/restore operation.
OpenEBS plugin supports incremental snapshots also. So, an entry can be added to this map which provides name of its parent backup.